File: tests/algorithms/transpiler/FS/matrix/count_paths.fs
// Generated 2025-08-17 13:19 +0700

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _idx (arr:'a array) (i:int) : 'a =
    if not (obj.ReferenceEquals(arr, null)) && i >= 0 && i < arr.Length then arr.[i] else Unchecked.defaultof<'a>
let rec _str v =
    match box v with
    | :? float as f -> sprintf "%.10g" f
    | _ ->
        let s = sprintf "%A" v
        s.Replace("[|", "[")
         .Replace("|]", "]")
         .Replace("; ", " ")
         .Replace(";", "")
         .Replace("\"", "")
let rec depth_first_search (grid: int array array) (row: int) (col: int) (visit: bool array array) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable grid = grid
    let mutable row = row
    let mutable col = col
    let mutable visit = visit
    try
        let row_length: int = Seq.length (grid)
        let col_length: int = Seq.length (_idx grid (int 0))
        if (((row < 0) || (col < 0)) || (row = row_length)) || (col = col_length) then
            __ret <- 0
            raise Return
        if _idx (_idx visit (int row)) (int col) then
            __ret <- 0
            raise Return
        if (_idx (_idx grid (int row)) (int col)) = 1 then
            __ret <- 0
            raise Return
        if (row = (row_length - 1)) && (col = (col_length - 1)) then
            __ret <- 1
            raise Return
        visit.[row].[col] <- true
        let mutable count: int = 0
        count <- count + (depth_first_search (grid) (row + 1) (col) (visit))
        count <- count + (depth_first_search (grid) (row - 1) (col) (visit))
        count <- count + (depth_first_search (grid) (row) (col + 1) (visit))
        count <- count + (depth_first_search (grid) (row) (col - 1) (visit))
        visit.[row].[col] <- false
        __ret <- count
        raise Return
        __ret
    with
        | Return -> __ret
and count_paths (grid: int array array) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable grid = grid
    try
        let rows: int = Seq.length (grid)
        let cols: int = Seq.length (_idx grid (int 0))
        let mutable visit: bool array array = Array.empty<bool array>
        let mutable i: int = 0
        while i < rows do
            let mutable row_visit: bool array = Array.empty<bool>
            let mutable j: int = 0
            while j < cols do
                row_visit <- Array.append row_visit [|false|]
                j <- j + 1
            visit <- Array.append visit [|row_visit|]
            i <- i + 1
        __ret <- depth_first_search (grid) (0) (0) (visit)
        raise Return
        __ret
    with
        | Return -> __ret
and main () =
    let mutable __ret : obj = Unchecked.defaultof<obj>
    try
        let __bench_start = _now()
        let __mem_start = System.GC.GetTotalMemory(true)
        let grid1: int array array = [|[|0; 0; 0; 0|]; [|1; 1; 0; 0|]; [|0; 0; 0; 1|]; [|0; 1; 0; 0|]|]
        ignore (printfn "%s" (_str (count_paths (grid1))))
        let grid2: int array array = [|[|0; 0; 0; 0; 0|]; [|0; 1; 1; 1; 0|]; [|0; 1; 1; 1; 0|]; [|0; 0; 0; 0; 0|]|]
        ignore (printfn "%s" (_str (count_paths (grid2))))
        let __bench_end = _now()
        let __mem_end = System.GC.GetTotalMemory(true)
        printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

        __ret
    with
        | Return -> __ret
ignore (main())

File: tests/algorithms/x/PHP/matrix/count_paths.php
<?php
error_reporting(E_ALL & ~E_DEPRECATED);
ini_set('memory_limit', '-1');
function _str($x) {
    if (is_array($x)) {
        $isList = array_keys($x) === range(0, count($x) - 1);
        if ($isList) {
            $parts = [];
            foreach ($x as $v) { $parts[] = _str($v); }
            return '[' . implode(' ', $parts) . ']';
        }
        $parts = [];
        foreach ($x as $k => $v) { $parts[] = _str($k) . ':' . _str($v); }
        return 'map[' . implode(' ', $parts) . ']';
    }
    if (is_bool($x)) return $x ? 'true' : 'false';
    if ($x === null) return 'null';
    return strval($x);
}
function _append($arr, $x) {
    $arr[] = $x;
    return $arr;
}
function depth_first_search($grid, $row, $col, &$visit) {
  $row_length = count($grid);
  $col_length = count($grid[0]);
  if ($row < 0 || $col < 0 || $row == $row_length || $col == $col_length) {
  return 0;
}
  if ($visit[$row][$col]) {
  return 0;
}
  if ($grid[$row][$col] == 1) {
  return 0;
}
  if ($row == $row_length - 1 && $col == $col_length - 1) {
  return 1;
}
  $visit[$row][$col] = true;
  $count = 0;
  $count = $count + depth_first_search($grid, $row + 1, $col, $visit);
  $count = $count + depth_first_search($grid, $row - 1, $col, $visit);
  $count = $count + depth_first_search($grid, $row, $col + 1, $visit);
  $count = $count + depth_first_search($grid, $row, $col - 1, $visit);
  $visit[$row][$col] = false;
  return $count;
}
function count_paths($grid) {
  $rows = count($grid);
  $cols = count($grid[0]);
  $visit = [];
  $i = 0;
  while ($i < $rows) {
  $row_visit = [];
  $j = 0;
  while ($j < $cols) {
  $row_visit = _append($row_visit, false);
  $j = $j + 1;
};
  $visit = _append($visit, $row_visit);
  $i = $i + 1;
};
  return depth_first_search($grid, 0, 0, $visit);
}
function main() {
  $grid1 = [[0, 0, 0, 0], [1, 1, 0, 0], [0, 0, 0, 1], [0, 1, 0, 0]];
  echo rtrim(_str(count_paths($grid1))), PHP_EOL;
  $grid2 = [[0, 0, 0, 0, 0], [0, 1, 1, 1, 0], [0, 1, 1, 1, 0], [0, 0, 0, 0, 0]];
  echo rtrim(_str(count_paths($grid2))), PHP_EOL;
}
main();
